spec.html

```diff
@@ -7652,19 +7650,44 @@ <h1>Jobs and Host Operations to Enqueue Jobs</h1>
+    <ul>
+      <li>GetActiveScriptOrModule() is _scriptOrModule_.</li>
+      <li>If _scriptOrModule_ is a Script Record or Module Record, let _ec_ be the topmost execution context on the execution context stack whose ScriptOrModule component is _scriptOrModule_. The Realm component of _ec_ is _scriptOrModule_.[[Realm]].</li>
+    </ul>
+
+    <p>At any particular time, an execution is <dfn id="job-preparedtoevaluatecode">prepared to evaluate ECMAScript code</dfn> if all of the following conditions are true:</p>
+    <ul>
+      <li>The execution context stack is not empty.</li>
+      <li>The Realm component of the topmost execution context on the execution context stack is a Realm Record.</li>
+    </ul>
+
+    <emu-note>
+      <p>Host environments may prepare an execution to evaluate code by pushing execution contexts onto the execution context stack. The specific steps are implementation-defined.</p>
+      <p>The specific choice of Realm is up to the host environment. This initial execution context and Realm is only in use before any callback function is invoked. When a callback function related to a Job, like a Promise handler, is invoked, the invocation pushes its own execution context and Realm.</p>
+    </emu-note>
+
+    <p>Particular kinds of Jobs have additional conformance requirements.</p>
+
     <emu-clause id="sec-hostenqueuepromisejob" aoid="HostEnqueuePromiseJob">
       <h1>HostEnqueuePromiseJob ( _job_, _realm_ )</h1>
       <p>HostEnqueuePromiseJob is a host-defined abstract operation that schedules the Job Abstract Closure _job_ to be performed, at some future time. The Abstract Closures used with this algorithm are intended to be related to the handling of Promises, or otherwise, to be scheduled with equal priority to Promise handling operations.</p>
-      <p>The _realm_ parameter is passed through to hosts with no normative requirements; it is either *null* or a Realm.</p>
+      <p>The _realm_ parameter is either *null* or a Realm Record.</p>
+
+      <p>The implementation of HostEnqueuePromiseJob must conform to the requirements in <emu-xref href="#sec-jobs"></emu-xref>. Additionally, Jobs enqueued by this hook must conform to all of the following requirements:</p>
+      <ul>
+        <li>If _realm_ is not *null*, each time _job_ is invoked the implementation must perform implementation-defined steps such that execution is <emu-xref href="#job-preparedtoevaluatecode">prepared to evaluate ECMAScript code</emu-xref> at the time of invocation.</li>
+        <li>Let _scriptOrModule_ be GetActiveScriptOrModule(). If _realm_ is not *null*, each time _job_ is invoked the implementation must perform implementation-defined steps such that _scriptOrModule_ is the <emu-xref href="#job-activescriptormodule">active script or module</emu-xref> at the time of invocation.</li>
```

Comment:
> Let `_scriptOrModule_` be `GetActiveScriptOrModule()`

It is still not entirely clear to me when this happens.

Reply:
It happens at the time when HostEnqueuePromiseJob is called. Is the confusion the sentence "Additionally, Jobs enqueued by this hook must conform to all of the following requirements:"?

Reply:
The confusion results from the combination of the two, yeah.

Perhaps

```suggestion
        <li>If _scriptOrModule_ is GetActiveScriptOrModule() at the time `HostEnqueuePromiseJob` is invoked, and if _realm_ is not *null*, each time _job_ is invoked the implementation must perform implementation-defined steps suc that _scriptOrModule_ is the <emu-xref href="#job-activescriptormodule">active script or module</emu-xref> at the time of invocation.</li>
```

This use of "if" to introduce names is kind of unusual outside of the math world, but I don't have a better way to put it.

Reply:
Yeah, not a fan of the "if" there. I don't think there's anything untoward with qualifying the "let" with "at the time `HostEnqueuePromiseJob` is invoked". I'll remove the confusing "Additionally, [...]" sentence.